File: crates/crystalline_cylinder_analysis/src/center_of_mass.rs
//! Streaming axial COM unwrapping and differentiation.

use crate::input::{resolve_shard_path, MappedShard, SafetensorView, TensorDtype};
use crate::model::{CaseSchema, ComSeries, DiscoveredDataset};
use ndarray::{Array1, ArrayView1, ArrayView3, Axis};

/// COM analysis controls shared by individual replicas.
#[derive(Clone, Copy, Debug)]
pub struct ComConfig {
    pub timestep: f64,
}

/// Stream one dataset and compute its unwrapped axial COM and velocity.
pub fn analyze_replica_com(dataset: &DiscoveredDataset, config: ComConfig) -> ComSeries {
    assert!(
        config.timestep.is_finite() && config.timestep > 0.0,
        "bad timestep"
    );
    let case = &dataset.manifest.case;
    let mut previous_wrapped = Array1::<f64>::zeros(case.n_particles);
    let mut unwrapped = Array1::<f64>::zeros(case.n_particles);
    let mut initialized = false;
    let mut steps = Vec::with_capacity(dataset.manifest.frame_count);
    let mut centers = Vec::with_capacity(dataset.manifest.frame_count);

    for shard in &dataset.manifest.shards {
        let path = resolve_shard_path(&dataset.manifest_path, &shard.file);
        let mapped = MappedShard::open(&path);
        let coordinate_name = coordinate_name(dataset.schema, &mapped);
        let coordinates = mapped.tensor(coordinate_name);
        let shard_steps = mapped.tensor("step");
        let frame_count = shard.frame_stop - shard.frame_start;
        validate_shapes(
            &path,
            coordinate_name,
            &coordinates,
            &shard_steps,
            frame_count,
            case.n_particles,
        );
        let shape = (frame_count, case.n_particles, coordinates.shape[2]);
        match coordinates.dtype {
            TensorDtype::F32 => process_shard(
                ArrayView3::from_shape(shape, coordinates.as_f32()).expect("bad shape"),
                &shard_steps,
                case.lx,
                &mut previous_wrapped,
                &mut unwrapped,
                &mut initialized,
                &mut steps,
                &mut centers,
            ),
            TensorDtype::F64 => process_shard(
                ArrayView3::from_shape(shape, coordinates.as_f64()).expect("bad shape"),
                &shard_steps,
                case.lx,
                &mut previous_wrapped,
                &mut unwrapped,
                &mut initialized,
                &mut steps,
                &mut centers,
            ),
            TensorDtype::Bool | TensorDtype::I8 | TensorDtype::I32 | TensorDtype::I64 => {
                panic!("bad dtype")
            }
        }
        // The mapping and all borrowed tensor views are dropped here.
    }

    assert_eq!(
        centers.len(),
        dataset.manifest.frame_count,
        "bad frame count"
    );
    assert!(centers.len() >= 2, "too few frames");
    let initial_step = steps[0] as f64;
    let elapsed_time: Vec<f64> = steps
        .iter()
        .map(|&step| (step as f64 - initial_step) * config.timestep)
        .collect();
    let velocity = finite_gradient(&centers, &elapsed_time);
    let zeros = vec![0.0; centers.len()];
    ComSeries {
        elapsed_time,
        x_center_mean: centers,
        x_center_std: zeros.clone(),
        x_velocity_mean: velocity,
        x_velocity_std: zeros,
        replicate_count: 1,
    }
}

/// Differentiate samples with NumPy-compatible endpoint behavior.
pub fn finite_gradient(values: &[f64], coordinates: &[f64]) -> Vec<f64> {
    assert_eq!(values.len(), coordinates.len(), "bad gradient");
    assert!(values.len() >= 2, "bad gradient");
    assert!(values.iter().all(|value| value.is_finite()), "bad values");
    assert!(
        coordinates.iter().all(|value| value.is_finite()),
        "bad time"
    );
    assert!(
        coordinates.windows(2).all(|pair| pair[1] > pair[0]),
        "bad time"
    );
    if values.len() == 2 {
        let slope = (values[1] - values[0]) / (coordinates[1] - coordinates[0]);
        return vec![slope, slope];
    }

    let count = values.len();
    let mut gradient = vec![0.0; count];
    for index in 1..count - 1 {
        let before = coordinates[index] - coordinates[index - 1];
        let after = coordinates[index + 1] - coordinates[index];
        let a = -after / (before * (before + after));
        let b = (after - before) / (before * after);
        let c = before / (after * (before + after));
        gradient[index] = a * values[index - 1] + b * values[index] + c * values[index + 1];
    }

    let first = coordinates[1] - coordinates[0];
    let second = coordinates[2] - coordinates[1];
    gradient[0] = -(2.0 * first + second) / (first * (first + second)) * values[0]
        + (first + second) / (first * second) * values[1]
        - first / (second * (first + second)) * values[2];

    let before = coordinates[count - 2] - coordinates[count - 3];
    let last = coordinates[count - 1] - coordinates[count - 2];
    gradient[count - 1] = last / (before * (before + last)) * values[count - 3]
        - (last + before) / (before * last) * values[count - 2]
        + (2.0 * last + before) / (last * (before + last)) * values[count - 1];
    assert!(
        gradient.iter().all(|value| value.is_finite()),
        "bad velocity"
    );
    gradient
}

fn coordinate_name(schema: CaseSchema, mapped: &MappedShard) -> &'static str {
    match schema {
        CaseSchema::BigLx if mapped.contains("coords") => "coords",
        CaseSchema::Confinement if mapped.contains("coords") => "coords",
        CaseSchema::Confinement if mapped.contains("position_cartesian") => "position_cartesian",
        _ => panic!("missing coords"),
    }
}

fn validate_shapes(
    path: &std::path::Path,
    coordinate_name: &str,
    coordinates: &SafetensorView<'_>,
    steps: &SafetensorView<'_>,
    frames: usize,
    particles: usize,
) {
    let _ = (path, coordinate_name);
    assert_eq!(coordinates.shape.len(), 3, "bad shape");
    assert_eq!(coordinates.shape[0], frames, "bad shape");
    assert_eq!(coordinates.shape[1], particles, "bad shape");
    assert_ne!(coordinates.shape[2], 0, "bad shape");
    assert_eq!(steps.shape, [frames], "bad shape");
    assert!(
        matches!(steps.dtype, TensorDtype::I32 | TensorDtype::I64),
        "bad dtype"
    );
}

fn step_at(steps: &SafetensorView<'_>, index: usize) -> i64 {
    match steps.dtype {
        TensorDtype::I32 => i64::from(steps.as_i32()[index]),
        TensorDtype::I64 => steps.as_i64()[index],
        TensorDtype::Bool | TensorDtype::F32 | TensorDtype::F64 | TensorDtype::I8 => {
            unreachable!("step dtype validated")
        }
    }
}

#[allow(clippy::too_many_arguments)]
fn process_shard<T: Copy + Into<f64>>(
    coordinates: ArrayView3<'_, T>,
    shard_steps: &SafetensorView<'_>,
    lx: f64,
    previous_wrapped: &mut Array1<f64>,
    unwrapped: &mut Array1<f64>,
    initialized: &mut bool,
    steps: &mut Vec<i64>,
    centers: &mut Vec<f64>,
) {
    for local_frame in 0..coordinates.len_of(Axis(0)) {
        let step = step_at(shard_steps, local_frame);
        assert!(
            steps.last().is_none_or(|previous| step > *previous),
            "bad steps"
        );
        let frame = coordinates.index_axis(Axis(0), local_frame);
        let axial = frame.index_axis(Axis(1), 0);
        centers.push(update_unwrapped_center(
            axial,
            lx,
            previous_wrapped,
            unwrapped,
            initialized,
        ));
        steps.push(step);
    }
}

fn update_unwrapped_center<T: Copy + Into<f64>>(
    axial: ArrayView1<'_, T>,
    lx: f64,
    previous_wrapped: &mut Array1<f64>,
    unwrapped: &mut Array1<f64>,
    initialized: &mut bool,
) -> f64 {
    assert_eq!(axial.len(), previous_wrapped.len(), "bad shape");
    let mut sum = 0.0;
    let mut compensation = 0.0;
    for ((&value, previous), accumulated) in axial
        .iter()
        .zip(previous_wrapped.iter_mut())
        .zip(unwrapped.iter_mut())
    {
        let wrapped = value.into();
        assert!(wrapped.is_finite(), "bad coords");
        if *initialized {
            let displacement = wrapped - *previous;
            *accumulated += displacement - lx * (displacement / lx).round();
        } else {
            *accumulated = wrapped;
        }
        *previous = wrapped;
        let corrected = *accumulated - compensation;
        let updated = sum + corrected;
        compensation = (updated - sum) - corrected;
        sum = updated;
    }
    *initialized = true;
    sum / previous_wrapped.len() as f64
}

File: packages/dynamics_analysis/src/dynamics/center_of_mass.zig
//! Streaming center-of-mass unwrapping and finite differences.

const std = @import("std");
const input = @import("../input/root.zig");
const schema = @import("../schema.zig");
const Options = @import("../options.zig").Options;
const ComSeries = @import("../result.zig").ComSeries;

pub const ComWorkspace = struct {
    previous_wrapped: []f64,
    unwrapped: []f64,
};

pub fn analyzeCenterOfMass(
    allocator: std.mem.Allocator,
    dataset: input.DatasetInput,
    options: Options,
) !ComSeries {
    if (dataset.shard_paths.len == 0) return error.NoInput;
    try schema.validateTimestep(options.timestep);

    var static_shard = try input.Shard.open(allocator, dataset.static.path);
    defer static_shard.deinit();
    const lx = try schema.readLx(try static_shard.tensor("lx"));

    var shard_set = try input.ShardSet.open(allocator, dataset.shard_paths);
    defer shard_set.deinit();

    const first_schema = try shard_set.shards[0].frameSchema();
    var total_frames: usize = 0;
    for (shard_set.shards) |*shard| {
        const candidate = try shard.frameSchema();
        try schema.requireCompatibleFrames(first_schema, candidate);
        total_frames = try std.math.add(usize, total_frames, candidate.frame_count);
    }

    const frame_range = try schema.resolveFrameRange(
        total_frames,
        options.frame_start,
        options.frame_stop,
        2,
    );
    const output_frames = frame_range.len;

    const centers = try allocator.alloc(f64, output_frames);
    errdefer allocator.free(centers);
    const steps = try allocator.alloc(i64, output_frames);
    defer allocator.free(steps);
    const previous_wrapped = try allocator.alloc(f64, first_schema.particle_count);
    defer allocator.free(previous_wrapped);
    const unwrapped = try allocator.alloc(f64, first_schema.particle_count);
    defer allocator.free(unwrapped);
    var workspace = ComWorkspace{
        .previous_wrapped = previous_wrapped,
        .unwrapped = unwrapped,
    };

    var initialized = false;
    var global_frame: usize = 0;
    var output_frame: usize = 0;
    outer: for (shard_set.shards) |*shard| {
        const frame_schema = try shard.frameSchema();
        const coordinates = try shard.tensor("coords");
        const shard_steps = try shard.tensor("step");
        for (0..frame_schema.frame_count) |local_frame| {
            defer global_frame += 1;
            if (global_frame < frame_range.start) continue;
            if (global_frame >= frame_range.stop) break :outer;

            const frame = try coordinates.frame(local_frame);
            centers[output_frame] = switch (frame.dtype) {
                .f32 => try unwrapFrame(
                    f32,
                    try frame.values(f32),
                    frame_schema.component_count,
                    lx,
                    &workspace,
                    initialized,
                ),
                .f64 => try unwrapFrame(
                    f64,
                    try frame.values(f64),
                    frame_schema.component_count,
                    lx,
                    &workspace,
                    initialized,
                ),
                else => return error.InvalidCoordinateDtype,
            };
            steps[output_frame] = try stepAt(shard_steps, local_frame);
            try schema.validateNextStep(
                if (output_frame > 0) steps[output_frame - 1] else null,
                steps[output_frame],
            );
            initialized = true;
            output_frame += 1;
        }
    }
    if (output_frame != output_frames) return error.FrameCountMismatch;

    const elapsed_time = try allocator.alloc(f64, output_frames);
    errdefer allocator.free(elapsed_time);
    const initial_step = steps[0];
    for (steps, elapsed_time) |step, *elapsed| {
        elapsed.* = @as(f64, @floatFromInt(step - initial_step)) * options.timestep;
    }
    const velocity = try finiteDifference(allocator, centers, elapsed_time);
    errdefer allocator.free(velocity);
    return .{
        .elapsed_time = elapsed_time,
        .center = centers,
        .velocity = velocity,
    };
}

pub fn finiteDifference(
    allocator: std.mem.Allocator,
    values: []const f64,
    coords: []const f64,
) ![]f64 {
    try schema.validateDerivativeInput(values, coords);

    const derivative = try allocator.alloc(f64, values.len);
    errdefer allocator.free(derivative);
    if (values.len == 2) {
        const slope = (values[1] - values[0]) / (coords[1] - coords[0]);
        derivative[0] = slope;
        derivative[1] = slope;
        return derivative;
    }

    for (1..values.len - 1) |index| {
        const before = coords[index] - coords[index - 1];
        const after = coords[index + 1] - coords[index];
        const a = -after / (before * (before + after));
        const b = (after - before) / (before * after);
        const c = before / (after * (before + after));
        derivative[index] = a * values[index - 1] + b * values[index] + c * values[index + 1];
    }

    const first = coords[1] - coords[0];
    const second = coords[2] - coords[1];
    derivative[0] = -(2.0 * first + second) / (first * (first + second)) * values[0] +
        (first + second) / (first * second) * values[1] -
        first / (second * (first + second)) * values[2];

    const count = values.len;
    const before = coords[count - 2] - coords[count - 3];
    const last = coords[count - 1] - coords[count - 2];
    derivative[count - 1] = last / (before * (before + last)) * values[count - 3] -
        (last + before) / (before * last) * values[count - 2] +
        (2.0 * last + before) / (last * (before + last)) * values[count - 1];
    return derivative;
}

fn stepAt(tensor: input.safetensors.TensorView, index: usize) !i64 {
    return switch (tensor.dtype) {
        .i32 => @as(i64, (try tensor.values(i32))[index]),
        .i64 => (try tensor.values(i64))[index],
        else => error.InvalidStepDtype,
    };
}

fn unwrapFrame(
    comptime T: type,
    values: []align(1) const T,
    component_count: usize,
    lx: f64,
    workspace: *ComWorkspace,
    initialized: bool,
) !f64 {
    try schema.validateCoordinateFrameLength(
        values.len,
        workspace.unwrapped.len,
        component_count,
    );
    var sum: f64 = 0.0;
    var compensation: f64 = 0.0;
    for (0..workspace.unwrapped.len) |particle| {
        const wrapped = @as(f64, @floatCast(values[particle * component_count]));
        if (!std.math.isFinite(wrapped)) return error.NonFiniteCoordinate;
        if (initialized) {
            const displacement = wrapped - workspace.previous_wrapped[particle];
            workspace.unwrapped[particle] += displacement - lx * @round(displacement / lx);
        } else {
            workspace.unwrapped[particle] = wrapped;
        }
        workspace.previous_wrapped[particle] = wrapped;

        const corrected = workspace.unwrapped[particle] - compensation;
        const updated = sum + corrected;
        compensation = (updated - sum) - corrected;
        sum = updated;
    }
    return sum / @as(f64, @floatFromInt(workspace.unwrapped.len));
}
